### DISCUSSION TITLE
RGB16 files for 8266 Universal Display

### DISCUSSION THREAD

**Discussion by swampdogmash** [arendst/Tasmota#17075] [Q&A]
Hello All

Does anyone know how to create the RGB16 format files that are mentioned in the Display pages [(here](https://tasmota.github.io/docs/Displays/#miscellaneous)) which can be used on non-esp32 devices with the command:

`Pfilename: = display an rgb 16-bit color (or jpg on ESP32) image when file system is present, Scripteditor contains a converter to convert jpg to special RGB16 pictures`

Scripteditor is mentioned and he approx of the format but it's not clear how this type of files would be created and I couldnt find a sensible reference on google for which script editor. Is there a more specific definition of these files and how they can be created please?

**Comment by barbudor** [comment]
Scripteditor is a specific tool for Tasmota scripting language that you will find reference here:
https://tasmota.github.io/docs/Scripting-Language/#optional-external-editor

**Reply by sfromis** [reply]
Are you saying that the Script editor can create those binary image files for displays?

**Reply by barbudor** [reply]
Not having tested but considering that @gemu2015 gave birth to both Scripting and Universal Display (if I'm correct) I would not be surprised that we are talking about the same Scripteditor...

**Reply by gemu2015** [reply]
script editor has a submenu where you find 3 additional tools.
1. font editor to edit a c source code font file directly to add or modify characters
2. converter to convert jpg pictures to RGB pictures. RGB pictures are in the 16 bit rgb format that a color display usually needs. thus no conversion and RAM is needed when drawing the picture. so it can be used on esp8266 or on esp32 without PSRAM
3. converter to convert Font h files to binary font files which can be used by display text to load additional file based fonts.

**Reply by swampdogmash** [reply]
Would it be possible to link this in the Displays page please?

**Reply by barbudor** [reply]
You are welcome to make edits on the documentation by clicking on the pencil icon on the top left of the page
All contributors are welcomed

**Reply by swampdogmash** [reply]
I tried, but wasn't sure how the "creates a new copy and needs you to notify a pull request" bit worked. Fairly new to Github. I'll give it a go.

**Comment by swampdogmash** [comment]
I managed to download the ScriptEditor.zip file and found that I couldn't get it to run on Windows 10. not sure why. The MAC version seemed to run ok. For anyone else wondering which submenu the converter is under, it's not as obvious as it might seem.

From the "Various" menu - swipe over "Start" and then select "Font Editor". Then select "convert picture to RGB16 format".

I am not sure what to do next though ?  Selecting a JPG file and then entering a size seemed to just create a script that looked as follows:

```
82,0000	128,0000
1,0000
final size: 82,0000	128,0000
```

@gemu2015  What should i do now?  Is a file created in the same directory as the source? - it looks like it but can you confirm pls?

**Reply by swampdogmash** [reply]
Using the file in the same folder as the source , uploaded to the File manager, I tried:

`displaytext [x0y0P/cal.rgb:]`

I keep getting the result :

`21:24:07.772 MQT: homeautomation/wemos/stat/ILI9488/RESULT = {"DisplayText":"[x0y0P/cal.rgb"}`

with no output on the screen. Am I doing something wrong? It looks like the last square bracket somehow gets lost. I've tried putting the same command in the display.bat file in the root of the filesystem too, which just resulted in a blank screen.
Is there a max file or pixel size for conversion either input or output? I am trying to convert a 480x320 JPEG to a 480x320 RGB file for use as the background of the LCD. It appears the image is being displayed, but is just all black.

**Reply by gemu2015** [reply]
yes the converted picture is stored in the same location as the source.

while rgb16 format should work with all color displays i realized that there is a strange issue with ili9488.

i currently can not explain what is wrong. i am very busy with other things so i can not investigate soon.

you should however get a slow version going when modifying the source file xdrv13_display.ino

locate these lines in code

  if (!strcmp(estr,"rgb")) {
    // special rgb format
    fp = ufsp->open(file, FS_FILE_READ);
    if (!fp) return;
    uint16_t xsize;
    fp.read((uint8_t*)&xsize, 2);
    uint16_t ysize;
    fp.read((uint8_t*)&ysize, 2);
#if 1
    renderer->setAddrWindow(xp, yp, xp + xsize, yp + ysize);
    uint16_t rgb[xsize];

and then modify the #if 1 to #if 0

**Comment by gemu2015** [comment]
ok i found the issue and it was related to a clash  between an ILI9488 and SD card.
this will not apply to your setup. without sd card everything works as expected.
so try with a smaller picture first.

**Reply by swampdogmash** [reply]
Interesting and thank you very much.  I noticed that for some reason:

The windows version of script editor didn't initially load for some reason but once I fully unpacked all the files in the entire ZIP it worked just fine to convert to RGB16. On windows the image i was converting displayed after the conversion process and the file copied over to the Tasmota instance and worked as expected.

On MAC Big sur, converting the image did result in "an" image popping up, but it was just black. I had assumed this wasn't representative of what was in the file, but it seems it was. This is then what was loading on the screen once the file was transferred to Tasmota - a black screen. This was based on used the same source JPEG file, just the different versions of the ScriptEditor relevant for the OS.

I now have a 480x320 RGB image able to load up on the screen - As you aren't able to explain the issue you mentioned above (and no need to currently) I am not sure what difference it would make to my situation - but thank you so much for looking at it.

I did notice that the Mac version of ScriptEditor was in English and the Windows one didn't seem to be. It also looked like the Windows version conversion tools didn't render the text on the buttons and 1/4 of the buttons were off the top of the screen with an unmoveable window...

**Reply by gemu2015** [reply]
i fixed the windows version text issue. was a missing font. (updated on dropbox)
i realized that the mac converter version stopped working with big sur. it still worked with catalina
i will try to fix the mac converter soon.
the script editor itself works up to the latest mac version ventura

**Reply by swampdogmash** [reply]
Ok, even if you don't get round to it straight away at least this is a record of some info on how to do things. I really struggled to find any info so Thank YOU.

**Comment by jkwim** [comment]
Rather than opening a new thread I thought piggy backing on this old thread.

I have a situation:

1. Using uDisplay driver with ST7789 based display
2. Converted following image using Script Editor

https://github.com/Bodmer/OpenWeather/blob/main/examples/TFT_eSPI_OpenWeather_LittleFS/data/icon50/cloudy.bmp

https://raw.githubusercontent.com/Bodmer/OpenWeather/main/examples/TFT_eSPI_OpenWeather_LittleFS/data/icon50/cloudy.bmp

![image](https://github.com/user-attachments/assets/881338ae-145c-4705-a4c6-611447629d71)

3. When the image gets displayed the colors seem to be inverted ie. White background instead of black background.

What could be the issue?

**Reply by gemu2015** [reply]
did you check with default colors
```
displaytext [f1s2x250y400Ci2]Red
displaytext [x250y430Ci3]Green
displaytext [x250y460Ci4]Blue
```
adapt x and y for your display

if these are OK, also the picture should be OK, else you must invert the display colors in descriptor


the cloudy.jpg picture is ok on my displays

**Reply by jkwim** [reply]
Ah! You did confirm the hunch that I had that the color mapping is wrong.

`backlog displaytext [zf1s2x150y100Ci2]Red; displaytext [s2x150y125Ci3]Green; displaytext [s2x150y150Ci4]Blue
`
shows
Red -> in Cyan
Green -> Pink
Blue -> Yellow

How do you change that in Display Descriptor/Display.ini?

```
:H,ST7789,240,240,16,SPI,1,*,*,*,*,*,*,*,40
:S,1,2,0,1,80,30
:I
01,A0
11,A0
3A,81,55
36,81,00
21,80
13,80
29,A0
:o,28
:O,29
:A,2A,2B,2C
:R,36
:0,C0,00,50,00
:1,A0,50,00,01
:2,00,00,00,02
:3,60,00,00,03
:i,20,21

```

**Reply by jkwim** [reply]
Manage to do a comparison with code but I cannot figure out the value that need to be there if this is the place to change the color mapping.

```
3A,81,55
36,81,00
```


```
#define ST7789_COLMOD 0x3A


    BEGIN_WRITE,
    WRITE_C8_D8, ST7789_COLMOD, 0x55, // 3: Set color mode, 16-bit color
    WRITE_C8_D8, 0x36, 0x00,

```

https://github.com/SuGlider/Arduino_GFX/blob/c309dfe08493a5f97a465b0ebca766f7ecaebcff/src/display/Arduino_ST7789.h#L56

**Reply by gemu2015** [reply]
invert display in descriptor either
:i,20,21
or
:i,21,20

**Reply by jkwim** [reply]
`:i,20,21 -> :i,21,20` solved the problem. 

Thanks


I tried` 36,81,00->36,81,08` to change `RGB->BGR` it did. not work

In case anybody else needs the info:

```
#define ST7789_MADCTL  0x36
 
/**
 * Memory Data Access Control Register (0x36H)
 * MAP:     D7  D6  D5  D4  D3  D2  D1  D0
 * param:   MY  MX  MV  ML  RGB MH  -   -
 *
 */
 
/* Page Address Order ('0': Top to Bottom, '1': the opposite) */
#define ST7789_MADCTL_MY  0x80
/* Column Address Order ('0': Left to Right, '1': the opposite) */
#define ST7789_MADCTL_MX  0x40
/* Page/Column Order ('0' = Normal Mode, '1' = Reverse Mode) */
#define ST7789_MADCTL_MV  0x20
/* Line Address Order ('0' = LCD Refresh Top to Bottom, '1' = the opposite) */
#define ST7789_MADCTL_ML  0x10
/* RGB/BGR Order ('0' = RGB, '1' = BGR) */
#define ST7789_MADCTL_RGB 0x00
```

From https://community.st.com/t5/stm32-mcus-products/unable-to-display-anything-on-st7789-screen-via-nucleo/td-p/104603

**Comment by sfromis** [comment]
@gemu2015  Posting in this thread which is already about making the ScriptEditor work to convert image files to RGB format.

I tried to do this on an extra computer with Windows 10, current update level, and saw it crashing when trying to convert this simple PNG image (also when using a JPG image).
![tasmota-logo-rgb-32](https://github.com/user-attachments/assets/5daff4fb-29f0-4774-802e-c41b50231e7f)

Here are my steps:
1. Unpacked the downloaded zip
2. Drilled down to `Desktop\ScriptEditor\ScriptEditor\Deployment-windows\ScriptEditor.app\Contents\Windows`
3. Created a shortcut to `ScriptEditor.exe`, moved it to Desktop
4. Started the script editor via the shortcut
5. Selected `VarioPort`, `Start`, `Fonteditor`, `convert picture to rgb16 format`
6. Loaded the small 32x32 PNG test image above
7. Set the string `32x32` in "desired picturesize XS x YS" popup, instead of the much larger default value
8. Saw a bit of display updating (including image correctly visible)
9. And then the script editor windows gone after a couple of seconds

Using the Windows event viewer, Appliation logs, I found info about the crash. When loading, these messages, not the crash, but maybe interesting
```
The description for Event ID 1 from source ScriptEditor cannot be found. Either the component that raises this event is not installed on your local computer or the installation is corrupted. You can install or repair the component on the local computer.
If the event originated on another computer, the display information had to be saved with the event.
The following information was included with the event: 
"2024-09-21 15:33:38.066 ScriptEditor[2928:d5c] -[NSTextView setAlphaValue:] unimplemented in /Users/Shared/cocotron-gemu/AppKit/NSView.m at 446"
```

Next the "Application Error" crash message:
```
Faulting application name: ScriptEditor.exe, version: 0.0.0.0, time stamp: 0x65263473
Faulting module name: Foundation.1.0.dll, version: 0.0.0.0, time stamp: 0x588ef606
Exception code: 0xc0000005
Fault offset: 0x0007cbd1
Faulting process id: 0xb70
Faulting application start time: 0x01db0c2a458b5d46
Faulting application path: C:\Users\tinyq\Desktop\ScriptEditor\ScriptEditor\Deployment-windows\ScriptEditor.app\Contents\Windows\ScriptEditor.exe
Faulting module path: C:\Users\tinyq\Desktop\ScriptEditor\ScriptEditor\Deployment-windows\ScriptEditor.app\Contents\Windows\Foundation.1.0.dll
Report Id: 33282406-55a2-47ea-baa9-6e3bd88b25e2
Faulting package full name: 
Faulting package-relative application ID: 
```

Next an "Information" message
```
Fault bucket 1840591002455447578, type 1
Event Name: APPCRASH
Response: Not available
Cab Id: 0
Problem signature:
P1: ScriptEditor.exe
P2: 0.0.0.0
P3: 65263473
P4: Foundation.1.0.dll
P5: 0.0.0.0
P6: 588ef606
P7: c0000005
P8: 0007cbd1
P9: 
P10: 
Attached files:
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER6563.tmp.dmp
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER6620.tmp.WERInternalMetadata.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER6630.tmp.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER662E.tmp.csv
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER665E.tmp.txt 
These files may be available here:
\\?\C:\ProgramData\Microsoft\Windows\WER\ReportArchive\AppCrash_ScriptEditor.exe_5094c54cfaab0986343381363814c8fddc52d9_f2b078bd_460d7991-854d-4731-933a-49face439349 
Analysis symbol:
Rechecking for solution: 0
Report Id: 33282406-55a2-47ea-baa9-6e3bd88b25e2
Report Status: 268435456
Hashed bucket: 7cc82ee14e376c39698b17c3ce23181a
Cab Guid: 0
```

And a new "Application Error", similar to the first, but different exception code
```
Faulting application name: ScriptEditor.exe, version: 0.0.0.0, time stamp: 0x65263473
Faulting module name: Foundation.1.0.dll, version: 0.0.0.0, time stamp: 0x588ef606
Exception code: 0xc000041d
Fault offset: 0x0007cbd1
Faulting process id: 0xb70
Faulting application start time: 0x01db0c2a458b5d46
Faulting application path: C:\Users\tinyq\Desktop\ScriptEditor\ScriptEditor\Deployment-windows\ScriptEditor.app\Contents\Windows\ScriptEditor.exe
Faulting module path: C:\Users\tinyq\Desktop\ScriptEditor\ScriptEditor\Deployment-windows\ScriptEditor.app\Contents\Windows\Foundation.1.0.dll
Report Id: dd6039f0-b385-402b-b590-676bb0ae8b16
Faulting package full name: 
Faulting package-relative application ID: 
```

And last another "Information" message, basically a copy of the prior one.

Another observation was than when trying on a small laptop screen (instead of a bigger display), the ScriptEditor window was taller than available screen size, and the image converter window also. In both cases with top of window being above top of screen, making them hard to use. The usual `ctrl-space` for window menu did not work, otherwise, I could have pressed `M` to move the window by arrow keys.

**Reply by gemu2015** [reply]
ok its some time ago that i tested this on windows.
on my mac it works ok.  will check the pc version asap.

according to screen size.
find the font editor folder inside app/contents/Resources/fonteditor
and open the file fonteditor.tcl with script editor  itself.
then edit this line to any value you want.
SET SCREEN 1024 768  10 10 "file editor/converter" 
if you comment this line the full screen size is used.

**Reply by gemu2015** [reply]
on my windows 10 (parallels) it works also.
the windows cross compiler i generated this app with is extremely old originally Windows XP.  (cocotron mac)
probably try with windows compatibility modes.

**Reply by gemu2015** [reply]
just in case someone wants to write a more modern converter the file format is extremely simple.
just the raw RGB16 data that can be directly copied to the screen needing only one display line of RAM buffer.
the RGB data is preceded by X and Y size of the picture as uint16 values.

**Reply by sfromis** [reply]
Looks like part of the reason for the window placement weirdness is that I had scaled the display to 150%, due to a 1080x1920 14" screen resulting in fonts getting rather small; letter height 1 mm without the parts above or below the common lower-case size. Still crashing with Windows 8, 7 and XP compatibility modes.

Anyway, I whipped together a quick script to convert images from RGB888 to RGB565 16-bit. Less reason to dust off the the Windows machine....

**Reply by gemu2015** [reply]
just for completeness: i meanwhile testet with a windows 11 laptop and it works without any crash.

**Reply by swampdogmash** [reply]
Just came back to this thread after a long time and re-downloaded the script editor to a different (new to me) windows 10 box. Navigated to Scripteditor-> deployment-windows -> Scripteditor.app -> contents -> windows and ran scripteditor.exe. It initially popped up an approval windows for network connectivity but then nothing. Re-running provides nothing either. I clearly got it to run previously but i followed my own instructions and it didn't seem to work.

I've tried in various compatibility modes to no avail, I can see the program in the details pane of task manager, but no window... Then went back, quite the task in task manager. ran it again and it "worked" I managed to follow the same instructions but I am getting a similar thing to @sfromis , i get the conversion script output but then the window just closes and there is no outputfile in the course folder (either from BMP or JPG source, regardless of what i put in the x,y options box (i tried same dimensions, different). am I back to me doing something wrong or is there something odd going on here?

EDIT - for good luck i did try using the source folder as c:\temp instead of my desktop, and this seemed to work ok. For reference, the script editor unpacked zip is also on my desktop.

Now how would I use a JPG or compress this RGB image? I tried various JPG formats to no avail -any more clues here?

**Reply by gemu2015** [reply]
there are no spaces allowed in the file path , probably this was the issue.
why do you want to compress the rgb image ???   this format MUST be uncompressed because it is used on low RAM devices like esp8266 only just because it does not need to be decompressed. on an esp32 you should use the JPEG directly at least if you have PSRAM.

**Reply by huyphamnhu** [reply]
@sfromis Could you share your script to convert images from RGB888 to RGB565 16-bit please?

**Reply by huyphamnhu** [reply]
@gemu2015 Could you take a look at my python script please?
[Tasmota-RGB565-Converter](https://github.com/huyphamnhu/Tasmota-RGB565-Converter)
There must be something wrong with my bit shifting logic. The picture colors are displayed incorrectly. It's blue everywhere and I don't know where the problem is 😢

**Reply by joba-1** [reply]
try to explicitly convert variables red, green and blue to 16bit before line 30, so that the shifts don’t overflow.

**Reply by huyphamnhu** [reply]
@joba-1 You are Amazing !!!
